nixos/modules/security/tpm2.nix

```diff
@@ -0,0 +1,120 @@
+{ lib, pkgs, config, ... }:
+let
+  cfg = config.security.tpm2;
+
+  # This snippet is taken from tpm2-tss/dist/tpm-udev.rules, but modified to allow custom user/groups
+  # The idea is that the tssUser is allowed to acess the TPM and kernel TPM resource manager, while
+  # the tssGroup is only allowed to access the kernel resource manager
+  # Therefore, if either of the two are null, the respective part isn't generated
+  udevRules = tssUser: tssGroup: ''
+    ${lib.optionalString (tssUser != null) ''KERNEL=="tpm[0-9]*", MODE="0660", OWNER="${tssUser}"''}
+    ${lib.optionalString (tssUser != null || tssGroup != null)
+      ''KERNEL=="tpmrm[0-9]*", MODE="0660"''
+      + lib.optionalString (tssUser != null) '', OWNER="${tssUser}"''
+      + lib.optionalString (tssGroup != null) '', GROUP="${tssGroup}"''
+     }
+  '';
+
+in {
+  options.security.tpm2 = {
+    enable = lib.mkEnableOption "Trusted Platform Module 2 support";
+
+    tssUser = lib.mkOption {
+      description = "Name of the tpm device-owner and service user, set if applyUdevRules is set";
+      type = lib.types.nullOr lib.types.str;
+      default = "tss";
+    };
+
+    tssGroup = lib.mkOption {
+      description = "Group of the tpm kernel resource manager (tpmrm) device-group, set if applyUdevRules is set";
+      type = lib.types.nullOr lib.types.str;
+      default = "tss";
+    };
+
+    applyUdevRules = lib.mkOption {
+      description = ''
+        Whether to make the /dev/tpm[0-9] devices accessible by the tssUser, or
+        the /dev/tpmrm[0-9] by tssGroup respectively
+      '';
+      type = lib.types.bool;
+      default = true;
+    };
+
+    abrmd = lib.mkOption {
+      description = "TPM2 userspace resource manager service";
+      type = lib.types.submodule ({
+        options = {
+          enable = lib.mkEnableOption "Trusted Platform 2 userspace resource manager daemon";
+
+          user = lib.mkOption {
+            description = ''
+              User tpm2-abrmd should run under. This user should have access to
+              the /dev/tpm[0-9] device.
+            '';
+            type = lib.types.str;
+            default = config.security.tpm2.tssUser;
+          };
+
+          package = lib.mkOption {
+            description = "tpm2-abrmd package to use";
+            type = lib.types.package;
+            default = pkgs.tpm2-abrmd;
```

Comment:
This needs `defaultText = "pkgs.tmp2-abrmd"` as well to render correctly in the manual

Reply:
Will do.